File: sdk/@launchdarkly/flutter/packages/observability/ios/launchdarkly_flutter_observability/Sources/launchdarkly_flutter_observability/LDNativeApiImpl.swift
import Flutter
import Foundation
import LaunchDarkly
import LaunchDarklyObservability
import LaunchDarklySessionReplay

/// Pigeon `LDNativeApi` implementation. Ported (line-for-line where it makes
/// sense) from `ObservabilityBridge.swift` in
/// `sdk/@launchdarkly/mobile-dotnet/macios/native/LDObserve/Sources/ObservabilityBridge.swift`,
/// with the Pigeon-generated wire DTOs taking the place of the .NET-side
/// `ObjcObservabilityOptions` / `ObjcSessionReplayOptions` types.
final class LDNativeApiImpl: NSObject, LDNativeApi {
    private let captureChannel: FlutterMethodChannel

    init(captureChannel: FlutterMethodChannel) {
        self.captureChannel = captureChannel
        super.init()
    }

    func start(
        mobileKey: String,
        observability: LDObservabilityOptions,
        replay: LDSessionReplayOptions,
        observabilityVersion: String,
        completion: @escaping (Result<LDStartResult, Error>) -> Void
    ) {
        do {
            let config = try makeConfig(
                mobileKey: mobileKey,
                observability: observability,
                replay: replay,
                observabilityVersion: observabilityVersion
            )

            let context = try makeAnonymousContext()

            // Resolve the pigeon `start` only after `LDClient.start` has
            // completed. By then the `Observability` plugin has registered and
            // published its `ObservabilityService`, so the native tracer/logger
            // are non-null and `exportSpans`/`recordLog` will be honored.
            // Returning before completion would let the Dart side begin
            // exporting while the bridge still returns nil, dropping early
            // lifecycle and flag-evaluation telemetry.
            LDClient.start(
                config: config,
                context: context,
                startWaitSeconds: 15.0,
                completion: { _ in
                    completion(.success(LDStartResult(nativeVersion: observabilityVersion)))
                }
            )
        } catch {
            completion(.failure(error))
        }
    }

    // Re-creates each Dart span on the native tracer so the native pipeline
    // stamps `session.id` (and applies sampling/batching). Mirrors MAUI's
    // `TraceBuilderAdapter`.
    func exportSpans(spans: [LDSpanData]) throws {
        guard let tracer = ObjcLDObserveBridge.getObjcTracer() else { return }
        for span in spans {
            let builder = tracer.spanBuilder(
                name: span.name ?? "",
                startTime: span.startTimeSeconds ?? 0,
                traceId: span.traceId ?? "",
                spanId: span.spanId ?? "",
                parentSpanId: span.parentSpanId ?? ""
            )

            builder.setAttributes(cleanAttributes(span.attributes) as NSDictionary)

            for case let event? in span.events ?? [] {
                let name = event.name ?? ""
                let eventAttrs = cleanAttributes(event.attributes)
                if eventAttrs.isEmpty {
                    builder.addEvent(name: name)
                } else {
                    builder.addEvent(name: name, attributes: eventAttrs as NSDictionary)
                }
            }

            if let status = span.statusCode, status != 0 {
                builder.setStatus(code: Int(status))
            }

            builder.end(time: span.endTimeSeconds ?? span.startTimeSeconds ?? 0)
        }
    }

    // Forwards the log to the native customer logger so it is emitted as a real
    // `LogRecord` with `session.id` and trace/span correlation.
    func recordLog(log: LDLogRecord) throws {
        guard let logger = ObjcLDObserveBridge.getObjcLogger() else { return }
        logger.recordLog(
            message: log.message ?? "",
            severity: Int(log.severityNumber ?? 9),
            traceId: log.traceId,
            spanId: log.spanId,
            isInternal: false,
            attributes: cleanAttributes(log.attributes)
        )
    }

    // Forwards a custom track event to the native observability SDK. Native
    // `track` always broadcasts a Session Replay `Track` timeline event and, when
    // `analytics.trackEvents` is enabled, emits the `track` span. This is why the
    // Dart side routes mobile track here instead of through `exportSpans`: only
    // the native track path reaches Session Replay. `contextKeys` (the LD
    // evaluation context's kind -> key pairs) is applied to the span so it is
    // attributed to the same context the web SDK records.
    func track(key: String, data: [String: Any?]?, metricValue: Double?, contextKeys: [String: String]?) throws {
        ObjcLDObserveBridge.track(
            key: key,
            data: data.map { cleanAttributes($0) },
            metricValue: metricValue.map { NSNumber(value: $0) },
            contextKeys: contextKeys
        )
    }

    // Forwards an identify to the native observability SDK (which caches the
    // context keys so the manual track path is attributed to the active context)
    // and Session Replay (which records who the user is on the active recording).
    // The Dart side calls this from the LaunchDarkly client's `afterIdentify`
    // hook. Mirrors MAUI's `ObservabilityHook.AfterIdentify` /
    // `SessionReplayHook.AfterIdentify`.
    func identify(contextKeys: [String: String], canonicalKey: String, completed: Bool) throws {
        let keys = contextKeys as NSDictionary
        ObjcLDObserveBridge.getObservabilityHookProxy()?.afterIdentify(
            contextKeys: keys,
            canonicalKey: canonicalKey,
            completed: completed
        )
        LDReplay.shared.hookProxy?.afterIdentify(
            contextKeys: keys,
            canonicalKey: canonicalKey,
            completed: completed
        )
    }

    /// Drops `nil` values so the native bridge receives a `[String: Any]`.
    private func cleanAttributes(_ attributes: [String: Any?]?) -> [String: Any] {
        guard let attributes = attributes else { return [:] }
        var result = [String: Any](minimumCapacity: attributes.count)
        for (key, value) in attributes {
            if let value = value {
                result[key] = value
            }
        }
        return result
    }

    private func makeConfig(
        mobileKey: String,
        observability: LDObservabilityOptions,
        replay: LDSessionReplayOptions,
        observabilityVersion: String
    ) throws -> LDConfig {
        var config = LDConfig(
            mobileKey: mobileKey,
            autoEnvAttributes: .enabled
        )
        config.startOnline = false

        let crashReportingEnabled: Bool = observability.instrumentation?.crashReporting ?? true
        // The Flutter API exposes a single `analytics.taps` flag, but natively taps need two
        // switches: `instrumentation.userTaps` runs the tap-detection machinery, and
        // `analytics.taps` publishes each detected tap as a `click` span. We drive both from the
        // one Flutter flag so taps are detected (not just published) regardless of the native
        // `Instrumentation` defaults. `views` is Android-only and ignored here.
        let tapsEnabled: Bool = observability.analytics?.taps ?? true
        let trackEventsEnabled: Bool = observability.analytics?.trackEvents ?? true
        let appLifecycleEnabled: Bool = observability.analytics?.appLifecycle ?? true
        let appLaunchEnabled: Bool = observability.analytics?.appLaunch ?? true
        let launchTimesEnabled: Bool = observability.instrumentation?.launchTimes ?? true
        let metricsEnabled: Bool = observability.metricsEnabled ?? true

        let serviceName: String = observability.serviceName ?? Defaults.serviceName
        let serviceVersion: String = observability.serviceVersion ?? Defaults.serviceVersion
        let otlpEndpoint: String = observability.otlpEndpoint ?? Defaults.otlpEndpoint
        let backendUrl: String = observability.backendUrl ?? Defaults.backendUrl
        let resourceAttributes: [String: AttributeValue] =
            buildResourceAttributes(observability.attributes)
        let customHeaders: [String: String] = observability.customHeaders ?? [:]
        let sessionBackgroundTimeout: TimeInterval = observability.sessionBackgroundTimeoutMillis
            .map { TimeInterval($0) / 1000.0 } ?? 15 * 60
        let logsApiLevel: ObservabilityOptions.LogLevel = mapLogLevel(observability.logsApiLevel)

        let tracesApi = ObservabilityOptions.AppTracing(
            includeErrors: observability.traces?.includeErrors ?? true,
            includeSpans: observability.traces?.includeSpans ?? true
        )
        let metricsApi: ObservabilityOptions.AppMetrics = metricsEnabled ? .enabled : .disabled
        let crashReporting = ObservabilityOptions.CrashReporting(
            source: crashReportingEnabled ? .KSCrash : .none
        )
        let instrumentation = ObservabilityOptions.Instrumentation(
            urlSession: .disabled,
            userTaps: tapsEnabled ? .enabled : .disabled,
            memory: .disabled,
            memoryWarnings: .disabled,
            cpu: .disabled,
            launchTimes: launchTimesEnabled ? .enabled : .disabled
        )
        // Taps and track events are both published as spans via Analytics. Track
        // events additionally drive the Session Replay `Track` timeline event,
        // which the Dart side reaches by calling `track` on this bridge.
        let analytics = ObservabilityOptions.Analytics(
            taps: tapsEnabled ? .enabled : .disabled,
            trackEvents: trackEventsEnabled ? .enabled : .disabled,
            appLifecycle: appLifecycleEnabled ? .enabled : .disabled,
            appLaunch: appLaunchEnabled ? .enabled : .disabled
        )

        let observabilityOptions = ObservabilityOptions(
            serviceName: serviceName,
            serviceVersion: serviceVersion,
            otlpEndpoint: otlpEndpoint,
            backendUrl: backendUrl,
            resourceAttributes: resourceAttributes,
            customHeaders: customHeaders,
            sessionBackgroundTimeout: sessionBackgroundTimeout,
            logsApiLevel: logsApiLevel,
            tracesApi: tracesApi,
            metricsApi: metricsApi,
            crashReporting: crashReporting,
            instrumentation: instrumentation,
            analytics: analytics
        )
        let observabilityPlugin = Observability(options: observabilityOptions)
        observabilityPlugin.distroAttributes = [
            "telemetry.distro.name": "observability-flutter-ios",
            "telemetry.distro.version": observabilityVersion,
        ]

        let sessionReplayOptions = makeSessionReplayOptions(replay: replay)
        let privacy = replay.privacy
        let flutterCaptureService = FlutterImageCaptureService(
            channel: captureChannel,
            maskTextInputs: privacy?.maskTextInputs ?? true,
            maskLabels: privacy?.maskLabels ?? false,
            maskImages: privacy?.maskImages ?? false,
            maskWebViews: privacy?.maskWebViews ?? false,
            minimumAlpha: privacy?.minimumAlpha ?? 0.02
        )
        config.plugins = [
            observabilityPlugin,
            SessionReplay(
                options: sessionReplayOptions,
                imageCaptureService: flutterCaptureService
            ),
        ]

        return config
    }

    private func makeSessionReplayOptions(
        replay: LDSessionReplayOptions
    ) -> LaunchDarklySessionReplay.SessionReplayOptions {
        let privacy = replay.privacy

        return LaunchDarklySessionReplay.SessionReplayOptions(
            isEnabled: replay.isEnabled ?? true,
            privacy: .init(
                maskTextInputs: privacy?.maskTextInputs ?? true,
                maskWebViews: privacy?.maskWebViews ?? false,
                maskLabels: privacy?.maskLabels ?? false,
                maskImages: privacy?.maskImages ?? false
            ),
            frameRate: replay.frameRate ?? 1.0
        )
    }

    /// Maps the OpenTelemetry log-severity number sent across the bridge onto the
    /// native `ObservabilityOptions.LogLevel`. The `none` sentinel
    /// (`Int.max` / `0x7fffffff`) maps to `.none`; otherwise the severity number
    /// is the matching raw value. Defaults to `.info` when nil or unrecognized.
    private func mapLogLevel(_ severity: Int64?) -> ObservabilityOptions.LogLevel {
        guard let severity = severity else { return .info }
        if severity >= Int64(Int32.max) { return .none }
        return ObservabilityOptions.LogLevel(rawValue: Int(severity)) ?? .info
    }

    private func makeAnonymousContext() throws -> LDContext {
        var contextBuilder = LDContextBuilder(key: "flutter-user-key")
        contextBuilder.kind("user")
        switch contextBuilder.build() {
        case .success(let context):
            return context
        case .failure(let error):
            throw error
        }
    }

    private func buildResourceAttributes(_ source: [String: Any?]?) -> [String: AttributeValue] {
        guard let source = source, !source.isEmpty else { return [:] }
        var result = [String: AttributeValue](minimumCapacity: source.count)
        for (key, value) in source {
            guard let value = value else { continue }
            if let av = AttributeValue(value) {
                result[key] = av
            } else {
                result[key] = .string(String(describing: value))
            }
        }
        return result
    }

    private enum Defaults {
        static let serviceName = "observability-flutter"
        static let serviceVersion = "0.1.0"
        static let otlpEndpoint = "https://otel.observability.app.launchdarkly.com:4318"
        static let backendUrl = "https://pub.observability.app.launchdarkly.com"
    }
}

File: sdk/@launchdarkly/flutter/packages/observability/android/src/main/kotlin/com/launchdarkly/launchdarkly_flutter_observability/LDNativeApiImpl.kt
package com.launchdarkly.launchdarkly_flutter_observability

import android.app.Activity
import android.app.Application
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import com.launchdarkly.observability.api.ObservabilityOptions
import com.launchdarkly.observability.bridge.LDObserveBridge
import com.launchdarkly.observability.context.LDObserveContext
import com.launchdarkly.observability.replay.PrivacyProfile
import com.launchdarkly.observability.replay.ReplayOptions
import com.launchdarkly.observability.sdk.LDObserve
import com.launchdarkly.observability.sdk.LDReplay
import io.flutter.plugin.common.MethodChannel
import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.api.common.Attributes
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.minutes

/**
 * Pigeon [LDNativeApi] implementation. Ported (line-for-line where it makes
 * sense) from `ObservabilityBridge.kt` in
 * `sdk/@launchdarkly/mobile-dotnet/android/native/LDObserve/.../ObservabilityBridge.kt`,
 * with the Pigeon-generated wire DTOs taking the place of the .NET-side
 * `LDObservabilityOptions` / `LDSessionReplayOptions` Java types.
 *
 * The current host [Activity] is set by the plugin via [ActivityAware] and
 * forwarded to `LDReplay.registerActivity(...)` after `LDObserve.init` so the
 * SDK has the activity it missed during `onActivityCreated`. This mirrors the
 * RN adapter at
 * `sdk/@launchdarkly/react-native-ld-session-replay/.../SessionReplayClientAdapter.kt`,
 * which has the same comment.
 */
internal class LDNativeApiImpl(
    private val application: Application,
    private val captureChannel: MethodChannel,
) : LDNativeApi {

    @Volatile
    var activity: Activity? = null

    override fun start(
        mobileKey: String,
        observability: LDObservabilityOptions,
        replay: LDSessionReplayOptions,
        observabilityVersion: String,
        callback: (Result<LDStartResult>) -> Unit,
    ) {
        // Mirrors the RN adapter: `LDObserve.init` (which runs OpenTelemetryRum.build())
        // requires the main thread, and consecutive starts get naturally serialized
        // when posted through the main looper.
        Handler(Looper.getMainLooper()).post {
            try {
                doStart(mobileKey, observability, replay, observabilityVersion)
                callback(Result.success(LDStartResult(nativeVersion = observabilityVersion)))
            } catch (t: Throwable) {
                callback(Result.failure(t))
            }
        }
    }

    private fun doStart(
        mobileKey: String,
        observability: LDObservabilityOptions,
        replay: LDSessionReplayOptions,
        observabilityVersion: String,
    ) {
        val resourceAttributes = buildResourceAttributes(
            observability.attributes,
            observabilityVersion,
        )

        // Drives both `analytics.taps` (publish `click` spans) and `instrumentations.userTaps`
        // (run tap detection: window wrapping + hit-testing) from the single Flutter flag.
        val tapsEnabled = observability.analytics?.taps ?: true

        val nativeObservabilityOptions = ObservabilityOptions(
            enabled = observability.isEnabled ?: true,
            serviceName = observability.serviceName ?: DEFAULT_SERVICE_NAME,
            serviceVersion = observability.serviceVersion ?: hostAppVersion(),
            contextFriendlyName = observability.contextFriendlyName,
            resourceAttributes = resourceAttributes,
            customHeaders = observability.customHeaders ?: emptyMap(),
            sessionBackgroundTimeout = observability.sessionBackgroundTimeoutMillis
                ?.milliseconds ?: 15.minutes,
            debug = false,
            otlpEndpoint = observability.otlpEndpoint ?: DEFAULT_OTLP_ENDPOINT,
            backendUrl = observability.backendUrl ?: DEFAULT_BACKEND_URL,
            logsApiLevel = mapLogLevel(observability.logsApiLevel?.toInt()),
            tracesApi = ObservabilityOptions.TracesApi(
                includeErrors = observability.traces?.includeErrors ?: true,
                includeSpans = observability.traces?.includeSpans ?: true,
            ),
            metricsApi = if (observability.metricsEnabled ?: true) {
                ObservabilityOptions.MetricsApi.enabled()
            } else {
                ObservabilityOptions.MetricsApi.disabled()
            },
            analytics = ObservabilityOptions.Analytics(
                taps = tapsEnabled,
                // The Flutter-facing `views` option maps to the native `screenViews`
                // (`screen_view` spans); the legacy OpenTelemetry activity-lifecycle spans it used
                // to control have been removed.
                screenViews = observability.analytics?.views ?: true,
                trackEvents = observability.analytics?.trackEvents ?: true,
                appLifecycle = observability.analytics?.appLifecycle ?: true,
                appLaunch = observability.analytics?.appLaunch ?: true,
            ),
            instrumentations = ObservabilityOptions.Instrumentations(
                crashReporting = observability.instrumentation?.crashReporting ?: true,
                launchTime = observability.instrumentation?.launchTimes ?: true,
                // The Flutter API exposes a single `analytics.taps` flag, but natively taps need two
                // switches: `instrumentation.userTaps` runs the tap-detection machinery (window
                // wrapping + hit-testing) and `analytics.taps` publishes each detected tap as a
                // `click` span. Drive both from the one Flutter flag - matching iOS - so disabling
                // taps stops detection too, instead of leaving the invasive capture running.
                userTaps = tapsEnabled,
                // Flutter navigation is reported explicitly through trackScreenView. Activity
                // lifecycle detection only sees the single host Activity and would overwrite the
                // active Flutter route after a foreground/resume or session reseed.
                screens = false,
            ),
        )

        val privacy = replay.privacy
        val maskTextInputs = privacy?.maskTextInputs ?: true
        // Resolve once so the Flutter capture resolution and the exporter scale
        // stay in sync (a mismatch produces oversized replay frames). Treat a
        // missing or non-positive scale as the 1.0 default; otherwise the exporter
        // would record an invalid scale while the Dart capture falls back to the
        // device pixel ratio, reintroducing the mismatch.
        val replayScale = replay.scale?.takeIf { it > 0 } ?: 1.0
        val nativeReplayOptions = ReplayOptions(
            enabled = replay.isEnabled ?: true,
            sampleRate = replay.sampleRate ?: 1.0,
            frameRate = replay.frameRate ?: 1.0,
            scale = replayScale,
            imageQuality = replay.imageQuality ?: 0.3,
            privacyProfile = PrivacyProfile(
                maskTextInputs = maskTextInputs,
                maskText = privacy?.maskLabels ?: false,
                maskImageViews = privacy?.maskImages ?: false,
                maskWebViews = privacy?.maskWebViews ?: false,
                minimumAlpha = (privacy?.minimumAlpha ?: 0.02).toFloat(),
            ),
        )

        val ldContext = LDObserveContext.builder(LDObserveContext.DEFAULT_KIND, "flutter-user-key")
            .anonymous(true)
            .build()

        LDObserve.init(
            application = application,
            mobileKey = mobileKey,
            ldContext = ldContext,
            observability = nativeObservabilityOptions,
            replay = nativeReplayOptions,
            imageCaptureService = FlutterImageCaptureService(
                channel = captureChannel,
                maskTextInputs = maskTextInputs,
                maskLabels = privacy?.maskLabels ?: false,
                maskImages = privacy?.maskImages ?: false,
                maskWebViews = privacy?.maskWebViews ?: false,
                minimumAlpha = privacy?.minimumAlpha ?: 0.02,
                scale = replayScale,
            ),
        )

        // The Flutter plugin attaches after the host Activity is already created,
        // so the LD session replay SDK missed onActivityCreated() for it. Manually
        // register it so the view tree starts being captured — without this, the
        // session replay shows a black screen.
        activity?.let { LDReplay.registerActivity(it) }
    }

    // Re-creates each Dart span on the native tracer so the native pipeline
    // stamps `session.id` (and applies sampling/batching). Mirrors MAUI's
    // `TraceBuilderAdapter`.
    override fun exportSpans(spans: List<LDSpanData>) {
        val tracer = LDObserveBridge.getKotlinTracer() ?: return
        for (span in spans) {
            val builder = tracer.spanBuilder(
                span.name ?: "",
                span.startTimeSeconds ?: 0.0,
                span.traceId ?: "",
                span.spanId ?: "",
                span.parentSpanId ?: "",
            )

            span.attributes?.let { builder.setAttributes(it) }

            span.events?.forEach { event ->
                if (event == null) return@forEach
                val name = event.name ?: ""
                val attrs = event.attributes
                if (attrs.isNullOrEmpty()) {
                    builder.addEvent(name)
                } else {
                    builder.addEvent(name, attrs)
                }
            }

            val status = span.statusCode?.toInt() ?: 0
            if (status != 0) {
                builder.setStatus(status)
            }

            builder.end(span.endTimeSeconds ?: span.startTimeSeconds ?: 0.0)
        }
    }

    // Forwards the log to the native customer logger so it is emitted as a real
    // `LogRecord` with `session.id` and trace/span correlation.
    override fun recordLog(log: LDLogRecord) {
        val logger = LDObserveBridge.getKotlinLogger() ?: return
        logger.recordLog(
            log.message ?: "",
            log.severityNumber?.toInt() ?: 9,
            log.traceId,
            log.spanId,
            false,
            log.attributes,
        )
    }

    // Forwards a custom track event to the native observability SDK. Native
    // `track` always broadcasts a Session Replay `Track` timeline event and, when
    // `analytics.trackEvents` is enabled, emits the `track` span. The Dart side
    // routes mobile track here instead of through `exportSpans` because only the
    // native track path reaches Session Replay. `contextKeys` (the LD evaluation
    // context's kind -> key pairs) is applied to the span so it is attributed to
    // the same context the web SDK records. The bridge owns the `Map` -> `LDValue`
    // conversion so this plugin stays independent of the LaunchDarkly
    // feature-flag SDK.
    override fun track(
        key: String,
        data: Map<String, Any?>?,
        metricValue: Double?,
        contextKeys: Map<String, String>?,
    ) {
        LDObserveBridge.track(key, data, metricValue, contextKeys)
    }

    // Forwards an identify to the native observability SDK (which caches the
    // context keys so the manual track path is attributed to the active context)
    // and Session Replay (which records who the user is on the active recording).
    // The Dart side calls this from the LaunchDarkly client's `afterIdentify`
    // hook. Reuses the same public cross-platform hook proxies that React Native
    // (`LDReplay.hookProxy`) and MAUI drive, mirroring the iOS plugin — so no
    // bespoke combined bridge entry point is needed.
    override fun identify(
        contextKeys: Map<String, String>,
        canonicalKey: String,
        completed: Boolean,
    ) {
        LDObserveBridge.getObservabilityHookProxy()
            ?.afterIdentify(contextKeys, canonicalKey, completed)
        LDReplay.hookProxy?.afterIdentify(contextKeys, canonicalKey, completed)
    }

    // Forwards a screen view to the native observability SDK, which emits the
    // `screen_view` span and broadcasts a Session Replay `Navigate` timeline
    // event. Native automatic screen detection (Activity lifecycle) never fires
    // for Flutter route changes inside the single host Activity, so the Dart side
    // reports them here (typically from a `NavigatorObserver`).
    override fun trackScreenView(
        name: String,
        screenClass: String?,
        screenId: String?,
        category: String?,
        properties: Map<String, Any?>?,
    ) {
        LDObserve.trackScreenView(name, screenClass, screenId, category, properties)
    }

    // Forwards a click to the native observability SDK, which emits the `click`
    // span and broadcasts a Session Replay `Click` timeline event. Flutter renders
    // its whole UI into one `FlutterSurfaceView`, so native hit-testing can only
    // ever name that view; the Dart side walks the widget tree instead and reports
    // the real target here.
    override fun trackClick(
        id: String?,
        tag: String?,
        classname: String?,
        text: String?,
        xpath: String?,
        screenId: String?,
        x: Long?,
        y: Long?,
        timestampMillis: Long?,
        properties: Map<String, Any?>?,
    ) {
        LDObserve.trackClick(
            id = id,
            tag = tag,
            classname = classname,
            text = text,
            xpath = xpath,
            screenId = screenId,
            x = x?.toInt(),
            y = y?.toInt(),
            timestampMillis = timestampMillis,
            properties = properties,
        )
    }

    // Declares that Dart now resolves clicks for the Flutter view, so native tap
    // detection stops reporting taps that land on it and each tap is counted once.
    // Handshaked from Dart rather than set at init: native starts before the widget
    // tree exists, so an app that never installs Dart click detection keeps the
    // coarse native clicks instead of silently reporting none.
    override fun setEmbedderClickHandling(enabled: Boolean) {
        LDObserve.setEmbedderClickHandling(enabled)
    }

    // Session Replay is the only native component with a teardown; the native
    // observability SDK keeps its automatic instrumentation running.
    override fun shutdown() {
        LDReplay.stop()
        LDReplay.flush()
    }

    /**
     * Maps the OpenTelemetry log-severity number sent across the bridge onto the
     * native [ObservabilityOptions.LogLevel]. Defaults to [ObservabilityOptions.LogLevel.INFO]
     * when null or unrecognized.
     */
    private fun mapLogLevel(severity: Int?): ObservabilityOptions.LogLevel {
        if (severity == null) return ObservabilityOptions.LogLevel.INFO
        return ObservabilityOptions.LogLevel.entries.firstOrNull { it.level == severity }
            ?: ObservabilityOptions.LogLevel.INFO
    }

    // The host app's `versionName`, which Flutter sets from `pubspec.yaml`.
    // Empty rather than a made-up version when the app declares none.
    private fun hostAppVersion(): String = try {
        @Suppress("DEPRECATION")
        application.packageManager
            .getPackageInfo(application.packageName, 0)
            .versionName
            .orEmpty()
    } catch (e: PackageManager.NameNotFoundException) {
        ""
    }

    private fun buildResourceAttributes(
        source: Map<String, Any?>?,
        observabilityVersion: String,
    ): Attributes {
        val builder = Attributes.builder()
        source?.forEach { (key, value) ->
            when (value) {
                is String -> builder.put(AttributeKey.stringKey(key), value)
                is Boolean -> builder.put(AttributeKey.booleanKey(key), value)
                is Long -> builder.put(AttributeKey.longKey(key), value)
                is Int -> builder.put(AttributeKey.longKey(key), value.toLong())
                is Double -> builder.put(AttributeKey.doubleKey(key), value)
                is Float -> builder.put(AttributeKey.doubleKey(key), value.toDouble())
                null -> Unit
                else -> builder.put(AttributeKey.stringKey(key), value.toString())
            }
        }
        builder.put("telemetry.distro.name", FLUTTER_DISTRO_NAME)
        builder.put("telemetry.distro.version", observabilityVersion)
        return builder.build()
    }

    companion object {
        private const val FLUTTER_DISTRO_NAME = "observability-flutter-android"
        private const val DEFAULT_SERVICE_NAME = "observability-flutter"
        private const val DEFAULT_OTLP_ENDPOINT =
            "https://otel.observability.app.launchdarkly.com:4318"
        private const val DEFAULT_BACKEND_URL =
            "https://pub.observability.app.launchdarkly.com"
    }
}
